executor/aggregate.go

```diff
@@ -93,7 +177,7 @@ func (e *HashAggExec) Next(goCtx goctx.Context) (Row, error) {
 	return retRow, nil
 }
 
-func (e *HashAggExec) getGroupKey(row Row) ([]byte, error) {
+func (e *HashAggExec) getGroupKey(row types.Row) ([]byte, error) {
```

Comment:
A lot of tests failed here because some errors happened during the execution of `codec.EncodeValue([]byte{}, vals...)`, the panic stack is:
```txt
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc4202e28f0, 0xc4212efe0f)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:408 +0x10e
panic(0x1c6d0e0, 0xc420379b40)
        /Users/jianzhang.zj/opt/go-1.9/src/runtime/panic.go:491 +0x283
github.com/pingcap/tidb/util/codec.EncodeDecimal(0xc42028bf00, 0x3, 0x8, 0x600040008, 0x0, 0x0, 0x0, 0x0, 0x1e2d940, 0xc420922000, ...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/util/codec/decimal.go:36 +0x4ef
github.com/pingcap/tidb/util/codec.encode(0x2872230, 0x0, 0x0, 0xc420e3a2c0, 0x1, 0x1, 0xc4212e0000, 0x1010fe8, 0x0, 0x1c72c00, ...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/util/codec/codec.go:99 +0x6d3
github.com/pingcap/tidb/util/codec.EncodeValue(0x2872230, 0x0, 0x0, 0xc420e3a2c0, 0x1, 0x1, 0x0, 0x0, 0x1e2d940, 0xc420922000, ...)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/util/codec/codec.go:172 +0x74
github.com/pingcap/tidb/executor.(*HashAggExec).getGroupKey(0xc4207ca4b0, 0x260dde0, 0xc420379b10, 0xc420379b10, 0x0, 0x0, 0x2b71d90, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/executor/aggregate.go:189 +0x34f
github.com/pingcap/tidb/executor.(*HashAggExec).execute(0xc4207ca4b0, 0x2c62d88, 0xc4202db880, 0x20002bec1b0, 0x20000000008)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/executor/aggregate.go:127 +0x1fc
github.com/pingcap/tidb/executor.(*HashAggExec).NextChunk(0xc4207ca4b0, 0x2c62d88, 0xc4202db880, 0xc4202c5980, 0x1, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/executor/aggregate.go:78 +0x411
github.com/pingcap/tidb/executor.(*recordSet).NextChunk(0xc420303200, 0x2c62d88, 0xc4202db880, 0xc4202c5980, 0xc420303200, 0xc4202c5980)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/executor/adapter.go:103 +0x56
github.com/pingcap/tidb/server.(*tidbResultSet).NextChunk(0xc4208a24b0, 0x2c62d88, 0xc4202db880, 0xc4202c5980, 0x4, 0x400)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/driver_tidb.go:302 +0x51
github.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc4202e28f0, 0x2c62d88, 0xc4202db880, 0x25f3a80, 0xc4208a24b0, 0x0, 0xc4208a24b0, 0x1)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:936 +0x3ef
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc4202e28f0, 0x2c62d88, 0xc4202db880, 0x25f3a80, 0xc4208a24b0, 0xc420370000, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:861 +0x668
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc4202e28f0, 0x2c62d88, 0xc4202db880, 0xc42079e721, 0x26, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:810 +0x115
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc4202e28f0, 0xc42079e721, 0x27, 0x27, 0x0, 0x0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:592 +0x595
github.com/pingcap/tidb/server.(*clientConn).Run(0xc4202e28f0)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/conn.go:452 +0x1b8
github.com/pingcap/tidb/server.(*Server).onConn(0xc4208793e0, 0x25f6600, 0xc4206c0908)
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/server.go:316 +0x144
created by github.com/pingcap/tidb/server.(*Server).Run
        /Users/jianzhang.zj/opt/gopath/src/github.com/pingcap/tidb/server/server.go:262 +0x496
```


`EncodeDecimal` will check `length` and `fraction` of the Datum parameter, throw a panic if some checks failed.

Why master branch works well ? It's because the different behavior of function `GetDatum()` between `types.DatumRow` and `chunk.Row`: `chunk.Row.GetDatum()` writes the field type's `Flen` and `Decimal` as the returned Datum's `Length` and `Fraction` while `types.DatumRow` not.

A workaround is:
```go
v, err := item.Eval(row)
...
if item.GetType().Tp == mysql.TypeNewDecimal {
    v.SetLength(0)
}
```
 @spongedu @coocood PTAL

Reply:
@zz-jason this workaround works for this condition. But I think it's not a radical solution.  I think we should normalize  the behavior of`GetDatum`  for `chunk.Row` and `types.DatumRow`, or we may have trouble somewhere else, just like I met in #5242 . Maybe we should open an issue for this and fix later?